### PR TITLE
One shared guard for Standard Schema, one isThenable, and a rule to keep them that way

### LINKED ISSUE #545: fix(core): thenable-returning Standard Schema validators bypass validation
## Summary

Schema validation uses `instanceof Promise` to decide whether to await a Standard Schema result. `await` duck-types `.then`; `instanceof Promise` tests prototype identity in the current realm. When a schema's `validate()` returns a **thenable that is not a native same-realm `Promise`**, the check falls through to the synchronous arm, reads `.issues` off the unresolved thenable (`undefined`), and reports success.

**Invalid input passes validation and the raw body continues down the pipeline.** No throw, no log, no error event.

This sits on the route `input` / `parse` path, which [`.standards/security.md`](.standards/security.md) treats as a trust boundary.

## Reproduction

Against `pipeline/validation.ts` directly, on a clean tree:

```ts
import { validateAgainst } from "./packages/routecraft/src/pipeline/validation.ts";

// Legal Standard Schema: the spec types validate() as Result | Promise<Result>,
// and `await` accepts any thenable.
const schema = {
  "~standard": {
    version: 1,
    vendor: "custom",
    validate() {
      return {
        then(resolve) {
          resolve({ issues: [{ message: "value must be a number" }] });
        },
      };
    },
  },
};

const res = await validateAgainst(schema, "not-a-number");
console.log(res);
```

```
result: {"ok":true,"value":"not-a-number"}
>>> BUG: invalid input PASSED validation, body flows downstream
```

The schema rejected the value. Validation reported `ok: true`.

## Affected sites

`packages/routecraft`:

- `src/pipeline/validation.ts:37` — `if (result instanceof Promise) result = await result;` (route input/output validation)
- `src/operations/validate.ts:86` — same shape, `.validate()` operation

Same pattern across the monorepo:

- `packages/ai/src/llm/enricher.ts:52`
- `packages/ai/src/llm/structured-output.ts:81`
- `packages/ai/src/mcp/validate-options.ts:292`
- `packages/testing/src/test-fn.ts:87`

## The correct check already exists in core

`src/event-bus.ts:135`:

```ts
if (result && typeof result.then === "function") {
```

Duck-typed, realm-safe, and measurably no slower than `instanceof` (73.2 ns/op vs 74.1 ns/op over 3M iterations).

## Evidence for the two failure modes

```
thenable    instanceof Promise : false   | await -> 42
cross-realm instanceof Promise : false   | await -> 7     (node:vm)
duck-typed  typeof v.then      : true    | true
```

## Reachability

Honest assessment: **narrow likelihood, high impact, trivial fix.**

Zod, valibot and arktype all return native promises, so mainstream usage is unaffected. It becomes reachable when:

- a **custom Standard Schema** implementation returns a thenable — explicitly permitted by the spec, which types the return as `Result | Promise<Result>` and says nothing about nativeness;
- a schema is constructed in **another realm** (`node:vm`, `worker_threads`, a separate bundled copy of a validation library), where even a genuine native `Promise` fails `instanceof`;
- a consumer **downlevels to ES5/ES2015** with a `Promise` polyfill, making every async validation non-native.

The narrowness is an argument about priority, not about whether to fix it. A validation bypass that fails open on a trust boundary should not be left in place when the fix is a one-line predicate change and the correct pattern already ships in the same package.

## Proposed fix

Replace the `instanceof Promise` predicate with a duck-typed thenable check at all six sites. Prefer a single shared helper in core over six copies, since this is the third distinct spelling of the idea in the codebase.

## Acceptance criteria

- [ ] Shared thenable helper in core (e.g. `isThenable(value): value is PromiseLike<unknown>`).
- [ ] `pipeline/validation.ts:37` and `operations/validate.ts:86` use it.
- [ ] The four sibling sites in `@routecraft/ai` and `@routecraft/testing` use it.
- [ ] `event-bus.ts:135` collapsed onto the same helper.
- [ ] Regression test: a Standard Schema returning a **non-native thenable** that reports issues must cause validation to fail, for both route `input` validation and the `.validate()` operation.
- [ ] Regression test: a cross-realm native promise (via `node:vm`) is awaited correctly.
- [ ] Consider an ESLint rule in `@routecraft/eslint-plugin-routecraft` banning `instanceof Promise` in favour of the helper, since this is the third independent occurrence.

## Note on test coverage

The full suite passes today with these defects present, because **no test anywhere in the repo uses a non-native thenable**. CI offers no protection here, so the regression tests above are the substance of the fix rather than a formality.

## Provenance

Found while adversarially reviewing #544, which proposed spreading this same `instanceof Promise` pattern to 20+ additional sites. That issue is closed; this is the real defect it surfaced.

### PR DESCRIPTION
**Pull request by ex0b1t**
Closes #545 and #575.

The thenable defect itself shipped in #660. This is the follow-up both tickets asked for: remove the duplication that let it spread to five sites, and add enforcement so it cannot grow back. Consolidation, plus three behaviour changes that fell out of it and are named below.

Four commits, in the order the work required: helpers first, adoption second, enforcement last. A lint rule added before the sites move fires on code that has not moved yet.

## 1. One `isThenable` (#545)

Three spellings survived #660: a private predicate in `pipeline/validation.ts`, a private copy at the AI SDK seam, and an inline unnamed duck-type in `event-bus.ts`. #545 names the event bus specifically.

`shared/thenable.ts` holds the one implementation, beside the existing `shared/standard-schema.ts`, and both core sites import it. It is **core-internal, not exported**: `DEFINITION_OF_DONE.md` bars new `@internal` symbols from a package's public entry point, and `index.ts:242` already records that rule being enforced once before. `@routecraft/ai` cannot cross that boundary and keeps its copy, now carrying the reason in its JSDoc. Two implementations with a stated reason rather than three without one.

**Behaviour change.** Migrating the event bus surfaced a bug on the line being migrated: the duck-type was followed by `result.catch(...)`, which a thenable does not carry, so the call threw from inside the surrounding `try` and the handler was logged as having thrown when it had returned normally. `Promise.resolve(result)` adapts it first. Two tests, both failing without the fix.

## 2. One `isStandardSchema` for seven guards (#575)

Seven boundaries hand-rolled the same "is this a Standard Schema with a callable validate" test, each with its own `~standard` index cast. They exist because `validateAgainst` dereferences that property unguarded, so the guard is the helper's precondition and belongs with the helper.

`isStandardSchema(value): value is StandardSchemaV1` is exported from core beside `formatSchemaIssues`. Public rather than `@internal`, because its consumers are in `@routecraft/ai` and `@routecraft/testing`, which cannot reach `packages/routecraft/src/shared/`. Built on the existing `standardExtensionOf` rather than re-deriving the extraction.

**Only the cast and the predicate moved.** Every caller keeps its own throw, error code and message, because they differ per boundary on purpose. A regression test per site pins each.

Two sites are not straight substitutions, in opposite directions:

- **`validateFnOptions` adopts partially.** It has two guards, not one: the first rejects a value that is not schema-shaped at all, the second a schema-shaped value with no validator. They diagnose different mistakes in different words, and collapsing them would lose a message. The first guard is not one of the duplicated spellings, so it stays and only the second adopts.
- **`agent()` collapses fully.** Its two guards already produced the identical message, so one predicate covers both.

`toAiSchema` keeps a cast for `~standard.jsonSchema`, the non-standard extension absent from the spec type. Only the `validate` half of its cast moved.

**Behaviour change: a callable schema now reads as a schema.** `standardExtensionOf` admitted only `typeof "object"`, so an ArkType schema, which is a function object carrying `~standard`, read as carrying no bag at all, while `validateAgainst` validates one happily. The test was narrower than the validation it feeds. Two consequences:

- ArkType is accepted wherever the guard runs, rather than refused by boundaries that previously indexed `~standard` directly.
- A suspension parked under an ArkType schema now hashes its rendered JSON Schema instead of falling back to vendor and version. That fallback is identical for every schema a vendor produces, so per `hash.ts`'s own comment the changed-schema half of the resume compatibility check could not fire for ArkType at all. It can now. **The digest changes for ArkType only**; Zod and the other object-shaped libraries are byte-identical. An ArkType suspension parked before this release fails its compatibility check once, on upgrade.

## 3. The cross-realm test (#545)

#545 documents two ways `instanceof Promise` misses an async schema and #660 tested one. The other is a genuine native `Promise` from another realm, which carries a different intrinsic. The fix already covered it, so this is the missing test rather than a missing fix, built with `node:vm` as the ticket specifies. It asserts its own premise (`instanceof` false, `toString` says `[object Promise]`) so it still means something if a runtime starts sharing intrinsics.

## 4. The lint rule, not where the ticket suggested (#545)

#545 suggests `@routecraft/eslint-plugin-routecraft`, and that is the one place it cannot go: `eslint.config.mjs` scopes that plugin to `examples/**/*.{js,mjs}`, so a rule there would never see `packages/*/src` and would have enforced nothing.

It follows the existing `.eslint-rules/test-case-doc.mjs` pattern instead, scoped to `packages/*/src/**`. The message gives the reason rather than the rule, and is worded to be actionable from any package, since `isThenable` is importable only from core.

Verified against a probe carrying `instanceof Promise` in a line comment, in JSDoc, and in a string literal: one report, on the expression. `globalThis.Promise` is caught too. There is no legitimate `instanceof Promise` anywhere in `packages/*/src`, so the rule starts from a clean tree with no exceptions.

`test-case-doc.mjs` ships with no tests of its own, so this adds none rather than inventing a harness for one rule.

## Review

`/dev:codereview` ran before this went up and found three things worth naming here, all fixed:

- **The ArkType regression above**, which I introduced. My original changeset claimed "no behaviour change" and was wrong.
- **A booby-trapped lint gate.** The new config block registered `custom` a second time. Two different objects under one plugin namespace is a config error, not a merge: `ConfigError: Cannot redefine plugin "custom"`. Latent only because no test currently lives under `packages/*/src`; the first person to colocate one would have aborted the entire `bun run lint` gate with an error naming neither their file nor this change. One shared `localRules` object now registers both repo-local rules.
- **The rule missed `globalThis.Promise`**, a member expression rather than an identifier.

## Deliberately not in scope

A fourth hand-rolled `~standard.jsonSchema` arm reader survives in `structured-output.ts`, with an untouched fifth in `mcp/server.ts`, and both have already drifted from `renderJsonSchemaArm`. Same shape as this work, but a new helper across four call sites with its own behaviour change, and both tickets scope this to the validate guard. Worth its own ticket.

## Verification

`bun run all` exit 0 at every commit. **2957 pass, 1 skip, 0 fail across 203 files**, against `cc652b9`'s baseline of 2942 / 1 / 0 across 200.

Changeset is `minor` for all three packages: `isStandardSchema` is new public API, and per `.standards/api-stability.md` breaking changes ship as `minor` in 0.x.

### CODE REVIEW

File: packages/ai/src/fn/fn.ts [hunk 1]
```diff
@@ -32,10 +32,7 @@ export function validateFnOptions(id: string, options: FnOptions): void {
       message: `agentPlugin: fn "${id}" "input" is required and must be a Standard Schema value (Zod/Valibot/ArkType/etc.).`,
     });
   }
-  const standard = (
-    options.input as { ["~standard"]?: { validate?: unknown } }
-  )["~standard"];
-  if (typeof standard?.validate !== "function") {
+  if (!isStandardSchema(options.input)) {
```
[the remark this answers is not in the document]
Comment (ex0b1t) [reply packages/ai/src/fn/fn.ts:38]: Verified and fixed in 29b050b.

Confirmed against the code before acting. `standardExtensionOf` was widened to accept callables, but this preliminary guard tests `options.input` independently and still rejected a function, so `validateFnOptions` was the one boundary left refusing ArkType:

```
fn/fn.ts       -> REJECTED: agentPlugin: fn "p" "input" is required and must be a Standard Schema value
mcp options    -> ACCEPTED
ai sdk bridge  -> ACCEPTED
testFn         -> ACCEPTED
```

Worse than a plain gap: the message it threw names ArkType among the libraries it claims to accept.

Applied the proposed widening, keeping the `~standard` object test intact (ArkType's bag is an object; only the outer value is callable). All four boundaries now accept it.

Also added a regression test asserting a callable schema is accepted at every guarded boundary in this package. It is hand-rolled as `Object.assign(() => undefined, { "~standard": {...} })` rather than importing ArkType, because `packages/ai` does not depend on it and the shape is what matters, not one library's implementation. It fails with the guard narrowed back.


---
_Generated by [Claude Code](https://claude.ai/code)_

File: .changeset/standard-schema-guard-consolidation.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
+---
+"@routecraft/routecraft": minor
+"@routecraft/testing": minor
+"@routecraft/ai": minor
+---
+
+One shared guard for "is this a Standard Schema", and one `isThenable` (#545, #575).
+
+Consolidation with no behaviour change. The thenable defect itself was fixed in #660; this is the follow-up that stops the duplication growing back.
```
[the remark this answers is not in the document]
Comment (ex0b1t) [reply .changeset/standard-schema-guard-consolidation.md]: Correct, and fixed in 29b050b.

The lead sentence contradicted the body: I added the ArkType and event-bus paragraphs after writing the opening and did not go back to it. It now reads "Mostly consolidation" and names all three behaviour changes up front: callable schemas accepted where they were refused, the ArkType suspension digest change, and the event-bus thenable handling.


---
_Generated by [Claude Code](https://claude.ai/code)_
